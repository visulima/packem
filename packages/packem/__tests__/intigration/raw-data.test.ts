import { rm } from "node:fs/promises";

import { readFileSync, writeFileSync } from "@visulima/fs";
import { temporaryDirectory } from "tempy";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import {
    createPackageJson,
    createPackemConfig,
    createTsConfig,
    execPackem,
    installPackage,
} from "../helpers";

describe("packem raw data", () => {
    let temporaryDirectoryPath: string;

    beforeEach(async () => {
        temporaryDirectoryPath = temporaryDirectory();

        await createPackemConfig(temporaryDirectoryPath);
    });

    afterEach(async () => {
        await rm(temporaryDirectoryPath, { recursive: true });
    });

    it("should generate js files with included raw content", async () => {
        expect.assertions(4);

        writeFileSync(
            `${temporaryDirectoryPath}/src/content.txt`,
            `thisismydata`,
        );
        writeFileSync(
            `${temporaryDirectoryPath}/src/index.ts`,
            `import content from './content.txt';

export const data = content;`,
        );

        await installPackage(temporaryDirectoryPath, "typescript");
        await createTsConfig(temporaryDirectoryPath);
        await createPackageJson(temporaryDirectoryPath, {
            devDependencies: {
                typescript: "*",
            },
            main: "./dist/index.cjs",
            module: "./dist/index.mjs",
        });

        const binProcess = await execPackem("build", [], {
            cwd: temporaryDirectoryPath,
            reject: false,
        });

        expect(binProcess.stderr).toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.mjs`,
        );

        expect(mjsContent).toBe(`const data$1 = "thisismydata";

const data = data$1;

export { data };
`);

        const cjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.cjs`,
        );

        expect(cjsContent).toBe(`'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const data$1 = "thisismydata";

const data = data$1;

exports.data = data;
`);
    });

    it("should generate js files with included raw content when the '?raw' query param is used", async () => {
        expect.assertions(4);

        writeFileSync(
            `${temporaryDirectoryPath}/src/content.txt`,
            `thisismydata`,
        );
        writeFileSync(
            `${temporaryDirectoryPath}/src/index.ts`,
            `import content from './content.txt?raw';

export const data = content;`,
        );

        await installPackage(temporaryDirectoryPath, "typescript");
        await createTsConfig(temporaryDirectoryPath);
        await createPackageJson(temporaryDirectoryPath, {
            devDependencies: {
                typescript: "*",
            },
            main: "./dist/index.cjs",
            module: "./dist/index.mjs",
        });

        const binProcess = await execPackem("build", [], {
            cwd: temporaryDirectoryPath,
            reject: false,
        });

        expect(binProcess.stderr).toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.mjs`,
        );

        expect(mjsContent).toBe(`const data$1 = "thisismydata";

const data = data$1;

export { data };
`);

        const cjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.cjs`,
        );

        expect(cjsContent).toBe(`'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const data$1 = "thisismydata";

const data = data$1;

exports.data = data;
`);
    });

    it("should generate js files with included raw content when importing .js?raw", async () => {
        expect.assertions(4);

        writeFileSync(
            `${temporaryDirectoryPath}/src/data.js`,
            `const message = "Hello from JS file";
console.log(message);
export default message;`,
        );
        writeFileSync(
            `${temporaryDirectoryPath}/src/index.ts`,
            `import jsContent from './data.js?raw';

export const data = jsContent;`,
        );

        await installPackage(temporaryDirectoryPath, "typescript");
        await createTsConfig(temporaryDirectoryPath);
        await createPackageJson(temporaryDirectoryPath, {
            devDependencies: {
                typescript: "*",
            },
            main: "./dist/index.cjs",
            module: "./dist/index.mjs",
        });

        // Configure packem to include .js files for raw imports
        await createPackemConfig(temporaryDirectoryPath);

        const binProcess = await execPackem("build", [], {
            cwd: temporaryDirectoryPath,
            reject: false,
        });

        expect(binProcess.stderr).toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.mjs`,
        );

        expect(mjsContent).toBe(`const data$1 = "const message = \\"Hello from JS file\\";\\nconsole.log(message);\\nexport default message;";

const data = data$1;

export { data };
`);

        const cjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.cjs`,
        );

        expect(cjsContent).toBe(`'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const data$1 = "const message = \\"Hello from JS file\\";\\nconsole.log(message);\\nexport default message;";

const data = data$1;

exports.data = data;
`);
    });

    it("should generate js files with included raw HTML content", async () => {
        expect.assertions(4);

        writeFileSync(
            `${temporaryDirectoryPath}/src/template.html`,
            `<!DOCTYPE html>
<html lang="en" data-theme="dark" class="no-js">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="Test HTML file with various attributes">
    <title data-testid="page-title">Test Page</title>
    <link rel="stylesheet" href="styles.css" crossorigin="anonymous" data-preload>
    <script src="app.js" defer async type="module" data-entry-point></script>
</head>
<body data-page="test" class="loading ready">
    <header role="banner" aria-label="Site header">
        <nav role="navigation" aria-labelledby="main-nav">
            <ul id="main-nav" class="nav-list">
                <li><a href="/" class="nav-link active" data-nav-item="home">Home</a></li>
                <li><a href="/about" class="nav-link" data-nav-item="about" aria-current="false">About</a></li>
            </ul>
        </nav>
    </header>
    <main role="main" id="content" data-content-area>
        <article class="post" data-post-id="123" itemscope itemtype="https://schema.org/Article">
            <h1 itemprop="headline" data-heading-level="1">Test Article</h1>
            <p class="excerpt" data-word-count="150">This is a test paragraph with <strong>bold text</strong> and <em>italic text</em>.</p>
            <img src="image.jpg" alt="Test image" width="800" height="600" loading="lazy" data-src="image.jpg">
        </article>
    </main>
    <footer role="contentinfo" data-footer>
        <p>&copy; 2024 Test Site</p>
    </footer>
</body>
</html>`,
        );
        writeFileSync(
            `${temporaryDirectoryPath}/src/index.ts`,
            `import htmlContent from './template.html?raw';

export const template = htmlContent;`,
        );

        await installPackage(temporaryDirectoryPath, "typescript");
        await createTsConfig(temporaryDirectoryPath);
        await createPackageJson(temporaryDirectoryPath, {
            devDependencies: {
                typescript: "*",
            },
            main: "./dist/index.cjs",
            module: "./dist/index.mjs",
        });

        const binProcess = await execPackem("build", [], {
            cwd: temporaryDirectoryPath,
            reject: false,
        });

        expect(binProcess.stderr).toBe("");
        expect(binProcess.exitCode).toBe(0);

        const mjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.mjs`,
        );

        expect(mjsContent).toMatchSnapshot("ESM output with raw HTML content");

        const cjsContent = readFileSync(
            `${temporaryDirectoryPath}/dist/index.cjs`,
        );

        expect(cjsContent).toMatchSnapshot("CommonJS output with raw HTML content");
    });
});
