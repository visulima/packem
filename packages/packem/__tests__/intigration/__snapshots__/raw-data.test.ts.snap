// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`packem raw data > should generate js files with included raw HTML content > CommonJS output with raw HTML content 1`] = `
"'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const data = "<!DOCTYPE html>\\n<html lang=\\"en\\" data-theme=\\"dark\\" class=\\"no-js\\">\\n<head>\\n    <meta charset=\\"UTF-8\\">\\n    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">\\n    <meta name=\\"description\\" content=\\"Test HTML file with various attributes\\">\\n    <title data-testid=\\"page-title\\">Test Page</title>\\n    <link rel=\\"stylesheet\\" href=\\"styles.css\\" crossorigin=\\"anonymous\\" data-preload>\\n    <script src=\\"app.js\\" defer async type=\\"module\\" data-entry-point></script>\\n</head>\\n<body data-page=\\"test\\" class=\\"loading ready\\">\\n    <header role=\\"banner\\" aria-label=\\"Site header\\">\\n        <nav role=\\"navigation\\" aria-labelledby=\\"main-nav\\">\\n            <ul id=\\"main-nav\\" class=\\"nav-list\\">\\n                <li><a href=\\"/\\" class=\\"nav-link active\\" data-nav-item=\\"home\\">Home</a></li>\\n                <li><a href=\\"/about\\" class=\\"nav-link\\" data-nav-item=\\"about\\" aria-current=\\"false\\">About</a></li>\\n            </ul>\\n        </nav>\\n    </header>\\n    <main role=\\"main\\" id=\\"content\\" data-content-area>\\n        <article class=\\"post\\" data-post-id=\\"123\\" itemscope itemtype=\\"https://schema.org/Article\\">\\n            <h1 itemprop=\\"headline\\" data-heading-level=\\"1\\">Test Article</h1>\\n            <p class=\\"excerpt\\" data-word-count=\\"150\\">This is a test paragraph with <strong>bold text</strong> and <em>italic text</em>.</p>\\n            <img src=\\"image.jpg\\" alt=\\"Test image\\" width=\\"800\\" height=\\"600\\" loading=\\"lazy\\" data-src=\\"image.jpg\\">\\n        </article>\\n    </main>\\n    <footer role=\\"contentinfo\\" data-footer>\\n        <p>&copy; 2024 Test Site</p>\\n    </footer>\\n</body>\\n</html>";

const template = data;

exports.template = template;
"
`;

exports[`packem raw data > should generate js files with included raw HTML content > ESM output with raw HTML content 1`] = `
"const data = "<!DOCTYPE html>\\n<html lang=\\"en\\" data-theme=\\"dark\\" class=\\"no-js\\">\\n<head>\\n    <meta charset=\\"UTF-8\\">\\n    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">\\n    <meta name=\\"description\\" content=\\"Test HTML file with various attributes\\">\\n    <title data-testid=\\"page-title\\">Test Page</title>\\n    <link rel=\\"stylesheet\\" href=\\"styles.css\\" crossorigin=\\"anonymous\\" data-preload>\\n    <script src=\\"app.js\\" defer async type=\\"module\\" data-entry-point></script>\\n</head>\\n<body data-page=\\"test\\" class=\\"loading ready\\">\\n    <header role=\\"banner\\" aria-label=\\"Site header\\">\\n        <nav role=\\"navigation\\" aria-labelledby=\\"main-nav\\">\\n            <ul id=\\"main-nav\\" class=\\"nav-list\\">\\n                <li><a href=\\"/\\" class=\\"nav-link active\\" data-nav-item=\\"home\\">Home</a></li>\\n                <li><a href=\\"/about\\" class=\\"nav-link\\" data-nav-item=\\"about\\" aria-current=\\"false\\">About</a></li>\\n            </ul>\\n        </nav>\\n    </header>\\n    <main role=\\"main\\" id=\\"content\\" data-content-area>\\n        <article class=\\"post\\" data-post-id=\\"123\\" itemscope itemtype=\\"https://schema.org/Article\\">\\n            <h1 itemprop=\\"headline\\" data-heading-level=\\"1\\">Test Article</h1>\\n            <p class=\\"excerpt\\" data-word-count=\\"150\\">This is a test paragraph with <strong>bold text</strong> and <em>italic text</em>.</p>\\n            <img src=\\"image.jpg\\" alt=\\"Test image\\" width=\\"800\\" height=\\"600\\" loading=\\"lazy\\" data-src=\\"image.jpg\\">\\n        </article>\\n    </main>\\n    <footer role=\\"contentinfo\\" data-footer>\\n        <p>&copy; 2024 Test Site</p>\\n    </footer>\\n</body>\\n</html>";

const template = data;

export { template };
"
`;
